### ISSUE TITLE
Wrong exit code 0 when unable to create files

### ISSUE THREAD

**Issue by l0b0** [schemaspy/schemaspy#331]
## Steps to Reproduce (for bugs)
1. Provoke the error "Unable to create directory /output/tables" (presumably by creating the directory mounted as /output as a different UID than the Docker image uses).

## Expected Behavior

Should result in a non-zero exit code.

## Current Behavior

Returns exit code 0.

## Context
* Database type: `pgsql`
* Full command including arguments how you run SchemaSpy: `command: ["-configFile", "/configuration/db.properties"]`

## Your Environment
* Version used: schemaspy/schemaspy:snapshot Docker image (6.0.0-rc2)
* Operating System and version: Arch Linux

**Comment by npetzall** [comment]
Are you doing a single schema analysis or multi?

**Comment by l0b0** [comment]
Single schema.

**Comment by npetzall** [comment]
Since System.exit is never called I'm assuming none of the exit codes work.
Need to verify if SpringApplication.exit calls System.exit

**Comment by Christian1984** [comment]
For me this issue still exists. If schemaspy isn't able to write a file, it shows ERROR - Failed to write file '<filename>' but exits with code 0. I am running schemaspy 6.0.0

**Comment by npetzall** [comment]
The error your getting is probably not fatal, you'll get some information. I suspect the diagrams are missing. 

Which could be a separate errorCode.

Just so that you can have gt or lt in scripts.

Am I wrong about the diagrams missing and the html is also missing there is a bigger issue.

**Comment by Christian1984** [comment]
Yes, sorry, I should have been more specific. Steps to reproduce:

- Let schemaspy run and output to ~\Desktop\out
- write-protect index.html in ~\Desktop\out
- Run schemaspy again

**Result:** 

Schemaspy prints Error Message "ERROR - Failed to write file index.html". It does not abort but continues to write all other files. Finally, it returns with status code 0!

**Comment by mprins** [comment]
since you say the program keeps running having the exit code of 0 makes sense because the program did not terminate abnormally; if you want a different exit(error) code returned the program must also terminate after the first error and generate the error code based on that event

**Comment by Christian1984** [comment]
I agree, @mprins ! But in that case the behaviour is not really consistent with the semantic of an ERROR, is it? I'd like to suggest one of the following two approaches:

a) If it is intended by design NOT to terminate the program after "Failed to write index.html", this should not be reported as an ERROR, but rather as a WARNING! 

or

b) If there is no good reason to keep the program running after "Failed to write index.html", this should indeed be handled as an ERROR and the application should immediately terminate with a code != 0.

I'd suggest option b). Since the output is basically useless without the presence of ALL output files, there is no good reason to keep the execution alive. I'd rather see the software terminate in that case!

What is the reason for the current behaviour? Why does the program continue to run after the above mentioned ERROR was raised? What is the rational behind this design?

**Comment by npetzall** [comment]
@Christian1984 could you share your semantics regarding ERROR?

I'm not gonna speak for all of us, but the software might have bugs, data might be corrupt, there might be errors in jdbc-drivers and so on and so on. I have avoided to make each ERROR fatal since extracting and getting some data is better than nothing. We are analyzing and trying to get information. This is also how I personally have used the software. So I would state the the claim that output is basically useless without the presence of ALL output files is wrong. Some people only want the html-files, some just want the diagrams.

But I'll be the first to admit that deciding if something should be INFO, WARNING, ERROR or FATAL is always tricky.

If schemaspy was only doing one thing, extracting one piece of data and it failed, that would a FATAL error.

But my personal look at log levels.
FATAL ERROR: Application terminates, the problem can't be solved.
ERROR: Application can still run, but the problem can't be handled or "healed"
WARNING: Something might go wrong, things look bad or we have tried to "heal"
INFO: Should more or less just inform about what we are going to do and what type of assumptions/configuration we are working with. Not log every single call/method or giving feedback that things are actually working as they should. For a cli it is somewhat okay to inform that we a progressing and not at a stand still but it should be modest.

But we are in control of the software so setting up a exit code schema isn't an issue.
0 no errors, all output. (no errors, no problems)
0-100 errors, some output (not fatal, will do as much work as possible)
>100 errors, no output (FATAL will exit when happens)

**Comment by Christian1984** [comment]
Hi Nils! I agree with what you stated above and I'll admit that my "output-is-useless"-statement was a bit harsh. It was heavily impacted by how I want to use the software! You guys are doing an amazing job and I'm sorry if my comment came out wrong.

Thanks for your suggestions. I like the ideas you described in the last paragraph: Having these different error codes would allow users to analyze and handle errors from batch scripts etc. quite nicely!
